be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp

```diff
@@ -70,10 +70,24 @@ bool SpillableHashJoinBuildOperator::need_input() const {
 Status SpillableHashJoinBuildOperator::set_finishing(RuntimeState* state) {
     auto defer_set_finishing = DeferOp([this]() { _join_builder->spill_channel()->set_finishing(); });
 
-    if (!_join_builder->spiller()->spilled()) {
+    if (spill_strategy() == spill::SpillStrategy::NO_SPILL ||
+        (!_join_builder->spiller()->spilled() &&
+         _join_builder->hash_join_builder()->hash_table().get_row_count() == 0)) {
         return HashJoinBuildOperator::set_finishing(state);
     }
 
+    // if this operator is changed to spill mode just before set_finishing,
+    // we should create spill task
+    if (!_join_builder->spiller()->spilled()) {
+        DCHECK(_is_first_time_spill);
+        _is_first_time_spill = false;
+        auto& ht = _join_builder->hash_join_builder()->hash_table();
+        RETURN_IF_ERROR(init_spiller_partitions(state, ht));
+
+        _hash_table_slice_iterator = _convert_hash_map_to_chunk();
+        RETURN_IF_ERROR(_join_builder->append_spill_task(state, _hash_table_slice_iterator));
```

Comment:
where can we known this change of spill mode? from profile or log?

Reply:
> where can we known this change of spill mode? from profile or log?

we can use SpillerMetrics in profile to check wether the spill mode changed, but the first spill  is triggered when set_finishing or push_chunk, the user does not need to perceive it, and it cannot be seen through the profile. 
I add some logs to observe it during my test.